src/fft.rg

```diff
@@ -172,105 +241,279 @@ function fft.generate_fft_interface(itype, dtype)
     return p
   end
 
-  local plan_dft = fftw_c["fftw_plan_dft_" .. dim .. "d"]
+  -- MAKE PLAN FUNCTIONS
 
+  -- Creates the FFT plan - GPU version. Called by make_plan if necessary.
   local make_plan_gpu
-  if use_cuda then
+  if gpu_available then
     __demand(__cuda, __leaf)
-    task make_plan_gpu(input : region(ispace(itype), dtype),
-                       output : region(ispace(itype), dtype),
-                       address_space : c.legion_address_space_t) : cufft_c.cufftHandle
-    where reads writes(input, output) do
+    task make_plan_gpu(input : region(ispace(itype), dtype_in),
+                       output : region(ispace(itype), dtype_out),
+                       plan : region(ispace(int1d), iface.plan),
+                       address_space : c.legion_address_space_t)
+    where reads writes(input, output, plan) do
+
+      var p = iface.get_plan(plan, true)
       var proc = get_executing_processor(__runtime())
-      if c.legion_processor_kind(proc) == c.TOC_PROC then
-        var i = c.legion_processor_address_space(proc)
-        regentlib.assert(address_space == i, "make_plan_gpu must be executed on a processor in the same address space")
-
-        var input_base = get_base(rect_t(input.ispace.bounds), __physical(input)[0], __fields(input)[0])
-        var output_base = get_base(rect_t(output.ispace.bounds), __physical(output)[0], __fields(output)[0])
-        var lo = input.ispace.bounds.lo:to_point()
-        var hi = input.ispace.bounds.hi:to_point()
-        var n : int[dim]
-        ;[data.range(dim):map(function(i) return rquote n[i] = hi.x[i] - lo.x[i] + 1 end end)]
-        var cufft_p : cufft_c.cufftHandle
-        var ok = cufft_c.cufftPlanMany(
-          &cufft_p,
-          dim,
-          &n[0],
-          [&int](0), 0, 0, -- inembed, istride, idist
-          [&int](0), 0, 0, -- onembed, ostride, odist
-          cufft_c.CUFFT_C2C,
-          1 -- batch
-        )
-        regentlib.assert(ok == cufft_c.CUFFT_SUCCESS, "cufftPlanMany failed")
-        return cufft_p
-      else
+      if c.legion_processor_kind(proc) ~= c.TOC_PROC then
         regentlib.assert(false, "make_plan_gpu must be executed on a GPU processor")
       end
+
+      var i = c.legion_processor_address_space(proc)
+      regentlib.assert(address_space == i, "make_plan_gpu must be executed on a processor in the same address space")
+
+      -- Get input and output bases
+      var input_base = get_base_in(rect_in_t(input.ispace.bounds), __physical(input)[0], __fields(input)[0]).base
+      var output_base = get_base_out(rect_out_t(output.ispace.bounds), __physical(output)[0], __fields(output)[0]).base
+      var lo = input.ispace.bounds.lo:to_point()
+      var hi = input.ispace.bounds.hi:to_point()
+      var n : int[dim] -- n is an array of size dim with the size of each dimension in the entries
```

Comment:
Most of this comment is transliterating the code. But what is `n`, really? It's the size of the input region? Something else?

Reply:
n is meant to be as follows: n[In] – Array of size rank, describing the size of each dimension, n[0] being the size of the outermost and n[rank-1] innermost (contiguous) dimension of a transform.

(from https://docs.nvidia.com/cuda/cufft/#cufftplanmany)